### DISCUSSION TITLE
onRecordsListRequest() isn't triggered when using $app.dao()

### DISCUSSION THREAD

**Discussion by sergio9929** [pocketbase/pocketbase#3123] [Q&A] [answered]
Hello,

I need to create a custom id for my `incidences` table that looks like this: 'A23-0' being [preffix][year]-[counter]
and the counter has to be reset every year.

To do this I have created a `counters` table which is updated every time I insert a new record to `incidences`. This works using `onRecordBeforeCreateRequest()`, but the problems come with the yearly reset.

I thought I could do something like this:
```js
onRecordsListRequest((e) => {
    const year = new DateTime().time().year()
    console.log('list', e.records)

    for (const record of e.records) {
        if (year != record.get('reset')) {
            record.set('reset', year)
            record.set('count', 0)
            $app.dao().saveRecord(record)
        }
    }
}, 'counters')
```
but it never triggers when I use `$app.dao().findFirstRecordByData()` (or similar) inside `onRecordBeforeCreateRequest()` and `onRecordAfterCreateRequest()`, only when I view `counters` table inside pocketbase's admin or my app.

My code:
```js
onRecordBeforeCreateRequest((e) => {
    const admin = e.httpContext.get('admin')
    // if (admin) return; // admins are allowed to do anything
    const authRecord = e.httpContext.get('authRecord')

    if (!admin || (admin && !e.record.get('key'))) {
        const counter = $app.dao().findFirstRecordByData('counters', 'collection', e.collection.name)

        let companyPreffix = ''
        if (authRecord) {
            const departmentsUsers = $app.dao().findFirstRecordByData('departmentsUsers', 'userId', authRecord?.id)
            $app.dao().expandRecord(departmentsUsers, ['departmentId.companyId'], null)
            companyPreffix = departmentsUsers.expandedOne('departmentId').expandedOne('companyId').get('preffix')
        }

        const keyParts = [
            companyPreffix?.toUpperCase(),
            counter.get('reset').toString().slice(-2),
            '-',
            counter.get('count')
        ]
        e.record.set('key', keyParts.join(''))
    }
    console.log('update', e.record)
}, 'incidences')

onRecordAfterCreateRequest((e) => {
    const record = $app.dao().findFirstRecordByData('counters', 'collection', e.collection.name)

    record.set('count', record.getInt('count') + 1)

    $app.dao().saveRecord(record)
}, 'incidences')
```

Is this the intended behavior and, if so, how can I achieve something similar?

**Comment by ganigeorgiev** [comment, marked as answer]
The `*Request` hooks are triggered **only** when the corresponding web API request action is performed.

If you want to listen to model/dao events you can attach to the [`OnModel*` hooks](https://pocketbase.io/docs/js-event-hooks/#db-hooks) but there is no hook for when the record is retrieved from the database (_`OnModelRead/Fetch` hook was requested before but I have some performance concerns of triggering the hook for each retrieval and for now it is left aside_).

I guess one workaround could be to have a [scheduled job](https://pocketbase.io/docs/js-jobs-scheduling/) that will take care to reset all counters at once, something like (_I haven't tested it_):

```js
// execute on January 1 for every year (UTC based)
cronAdd("resetCounters", "0 0 1 1 *", () => {
    try {
        const year = new DateTime().time().year()

        // run a raw query statement to update all records in a single run
        $app.dao().db().newQuery("UPDATE counters SET count=0, year={:year}")
            .bind({ 'year': year })
            .execute()
    } catch (err) {
        console.log("failed to reset counters:", err)
    }
})
```

**Reply by ganigeorgiev** [reply]
_One side note to keep in mind when dealing with manual counters is that they are prone to races and it could be better to wrap the read and write statements of the after record create counter in a transaction:_

```js
// using the onModel* hook so that the counter is updated even when
// a new record is created programmatically and not via the API
onModelAfterCreate((e) => {
    $app.dao().runInTransaction((txDao) => {
        const record = txDao.findFirstRecordByData('counters', 'collection', e.model.collection().name)

        record.set('count', record.getInt('count') + 1)

        txDao.saveRecord(record)
    });
}, 'incidences')
```
